### PR TITLE
fix: CsvDocument comma-ended lines and `Nullable<T>` parsing

### LINKED ISSUE #27: Comma-ended line causes IndexOutOfRangeException
### Description

Parsing comma-ended CSV line with CsvDocument throws an IndexOutOfRangeException, while Sep and CsvHelper do not.

### Reproduction Steps

```csharp
using Csv;

var source = """
id,name
1,Alice
2,
3,Charlie

"""u8.ToArray();

var document = CsvSerializer.ConvertToDocument(source);
foreach (var row in document.Rows)
{
    var id = row["id"].GetValue<int>();
    var name = row["name"].GetValue<string>();
    Console.WriteLine($"id: {id}, name: {name}");
}
```

### Expected

`row["name"].GetValue<string>();` returns null, without throwing Exception

### Actual

When reading `2,`, `IndexOutOfRangeException` was thrown.

```
Unhandled exception. System.IndexOutOfRangeException: Index was outside the bounds of the array.
```

### Version

* .NET SDK: 10.0.102
* CsvCSharp: 1.0.2

### LINKED ISSUE #28: Error when trying to parse CSV with Nullable<Int32> property
Hi @nuskey8,

I try to parse a CSV with the following "type":
``` csharp
public class Data
{
    public string Name { get; set; }
    public int? Age { get; set; }
}
```

And the following content: `"Matthias,40"`.
When I parse with the following code:
``` csharp
var options = new CsvOptions() {
    HasHeader = false
};
var data = CsvSerializer.Deserialize<Data>(System.Text.Encoding.UTF8.GetBytes("Matthias,40"), options);

```` 
I got this exception:
`System.ArgumentException : GenericArguments[0], 'System.Nullable`1[System.Int32]', on 'Csv.Formatters.NullableFormatter`1[T]' violates the constraint of type 'T'.
---- System.TypeLoadException : GenericArguments[0], 'System.Nullable`1[System.Int32]', on 'System.Nullable`1[T]' violates the constraint of type parameter 'T'.`

I take a look at the code, and found that in `src/Csv/Formatters/StandardFormatterProvider.cs` line 40, you create the `NullableFormater` with the nullable type instead of the underlying one:
`(ICsvFormatter<T>)Activator.CreateInstance(typeof(NullableFormatter<>).MakeGenericType(typeof(T).GenericTypeArguments[0]))!;`

### PR DESCRIPTION
**Pull request by paq**
- Fix comma-ended line (e.g, `2,` ) causing IndexOutOfRangeException in CsvDocument (#27)
  - Fix NullableFormatter construction failing on Nullable<T> properties (#28)
  - Register CharFormatter in StandardFormatterProvider
  - Reorganize CsvDocumentTests into `Test_CsvDocument_$Category_$TestCase` naming convention and add edge-case coverage

  Fixes #27
  Fixes #28

### CODE REVIEW

File: tests/Csv.Tests/CsvDocumentTests.cs [hunk 1]
```diff
@@ -57,7 +59,783 @@ public void Test_CsvDocument_Key()
...
+    [Test]
+    public void Test_CsvDocument_Quoting_EmptyQuotedString_IsNotNull()
+    {
+        var csv = "a,b,c\n,\"\",value"u8;
+
+        var document = CsvSerializer.ConvertToDocument(csv.ToArray());
+
+        Assert.That(document.Rows.Length, Is.EqualTo(1));
+        Assert.That(document.Rows[0]["a"].GetValue<string>(), Is.Null);
+        Assert.That(document.Rows[0]["b"].GetValue<string>(), Is.EqualTo(""));
```
Comment (paq) [inline tests/Csv.Tests/CsvDocumentTests.cs:433]: Should we treat a missing field and `""` as both null?

File: tests/Csv.Tests/CsvDocumentTests.cs [hunk 2]
```diff
@@ -57,7 +59,783 @@ public void Test_CsvDocument_Key()
...
+        // Trailing whitespace after the closing quote prevents the separator from
+        // being found, so the remainder of the line is not split correctly.
+        var csv = "Name,Age\n\"Alice\"  ,21\nBob,35"u8;
+
+        var document = CsvSerializer.ConvertToDocument(csv.ToArray());
+
+        Assert.That(document.Rows.Length, Is.EqualTo(3));
+        Assert.That(document.Rows[0].Length, Is.EqualTo(1));
+        Assert.That(document.Rows[0][0].GetValue<string>(), Is.EqualTo("Alice"));
+    }
```
Comment (paq) [inline tests/Csv.Tests/CsvDocumentTests.cs:422]: I don't understand the correct behavior for these cases. I'm treating the current state as correct

File: tests/Csv.Tests/CsvDocumentTests.cs (hunk 1 above)
Comment (nuskey8) [reply tests/Csv.Tests/CsvDocumentTests.cs:433]: As shown in the test case, I think it's fine to use `string.Empty` for explicit `\"\"` and null for whitespace.

**Comment by nuskey8** [comment]
Merged the PR. Thanks!
